MaxText/layers/linears.py

```diff
@@ -495,11 +495,29 @@ def load_balance_loss(self, top_k_indices, logits):
 
   def get_einsum(self, rhs_mesh_axes: Tuple[Optional[str], ...] = ()):
     if self.quant:
-      einsum_op = self.quant.einsum(rhs_mesh_axes)
+
+      def aqt_einsum(*args, **kwargs):
+        # aqt einsum doesn't support kwargs like precision
+        max_logging.log(f"[WARNING] {kwargs=} argument is not supported and will be ignored in aqt_einsum")
+        return self.quant.einsum(rhs_mesh_axes)(*args)
+
+      einsum_op = aqt_einsum
     else:
       einsum_op = jnp.einsum
     return einsum_op
 
+  def is_expert_parallelism_enabled(self):
+    return self.config.ici_expert_parallelism > 1 or self.config.dcn_expert_parallelism > 1
+
+  def maybe_all_gather_kernel_weight_in_expert_parallelism(self, kernel, kernel_axes):
+    if self.is_expert_parallelism_enabled():
```

Comment:
Is this necessary even with expert parallelism? Does the compiler not find the right solution without this sharding constraint?

Reply:
Yeap, I have tested and we need this line to let compiler understand the sharding.